invenio_records_resources/services/files/service.py

```diff
@@ -90,6 +94,18 @@ def init_files(self, identity, id_, data, uow=None):
         """Initialize the file upload for the record."""
         record = self._get_record(id_, identity, "create_files")
 
+        # All brand-new drafts don't allow exceeding files limit (while added via rest API).
+        # Old records that already had more files than limited can continue adding files.
+        # In case files amount goes back to under limit, users lose the privilege of adding more files.
+        resulting_files_count = record.files.count + 1
+        maxFiles = current_app.config.get("APP_RDM_DEPOSIT_FORM_QUOTA", {}).get(
+            "maxFiles", {}
+        )
+        if maxFiles and resulting_files_count == maxFiles + 1:
+            raise FilesCountExceededException(
```

Comment:
This will be visible in REST api, but in the UI as in the UI we have another check for this: https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/FileUploader/FileUploader.js#L87![Screenshot 2023-11-24 at 15 42 19](https://github.com/inveniosoftware/invenio-records-resources/assets/61321254/16ebc4f8-c664-4640-948c-3537822661f5)